src/python/pants/backend/experimental/jvm/target_types.py

```diff
@@ -0,0 +1,69 @@
+# Copyright 2020 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from __future__ import annotations
+
+import logging
+
+from pants.engine.target import (
+    COMMON_TARGET_FIELDS,
+    Dependencies,
+    Sources,
+    StringSequenceField,
+    Target,
+)
+
+logger = logging.getLogger(__name__)
+
+
+class JavaSources(Sources):
+    expected_file_extensions = (".java",)
+
+
+class JavaTestsSources(JavaSources):
+    default = ("*Test.java",)
+
+
+class JavaTests(Target):
+    alias = "java_tests"
+    core_fields = (
+        *COMMON_TARGET_FIELDS,
+        Dependencies,
+        JavaTestsSources,
+    )
+    help = "Java tests"
+
+
+class JavaLibrarySources(JavaSources):
+    default = ("*.java",) + tuple(f"!{pat}" for pat in JavaTestsSources.default)
+
+
+class JavaLibrary(Target):
+    alias = "java_library"
+    core_fields = (
+        *COMMON_TARGET_FIELDS,
+        Dependencies,
+        JavaLibrarySources,
+    )
+    help = "Java source code"
+
+
+class MavenRequirementsField(StringSequenceField):
+    alias = "maven_requirements"
+    help = (
+        "A sequence of Maven coordinate strings, e.g. ['org.scala-lang:scala-compiler:2.12.0', "
+        "'org.scala-lang:scala-library:2.12.0']."
+    )
+
+
+class JvmLockfileSources(Sources):
```

Comment:
It'd be good to override `help` for this field, including mentioning the intended semantics that it can be empty if the lockfile has not yet been generated, but then that you should update the sources field after generation.

Reply:
Done, PTAL at the new phrasing